File: eval-score/node/src/scorer.ts
import { EvalRow, EvaluatorName, MetricResult, ScoringResult } from './types';
import { evaluateRowAssertions } from './assertion-checker';
import { createJudge, Judge, metricFromJudge } from './judge-providers';
import { WorkIQClient } from './workiq-client';
import { DEFAULT_M365_EVALUATORS, deriveRowStatus, resolveRowEvaluators, setRowError } from './eval-document';
import { createThrottleGate } from './throttle-gate';

const DELAY_MS = 500;

function delay(ms: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, ms));
}

export async function scoreAnswers(
  rows: EvalRow[],
  client: WorkIQClient,
  options?: {
    tenantId?: string;
    judgeProvider?: 'workiq' | 'github-copilot' | 'azure-openai';
    fallbackJudgeProvider?: 'github-copilot' | 'azure-openai' | 'none';
    judge?: Judge;
    fallbackJudge?: Judge;
    judgeAgentId?: string;
    evaluators?: EvaluatorName[];
    concurrency?: number;
    delayMs?: number;
    threshold?: number;
    onProgress?: (completed: number, total: number) => void;
    onRowComplete?: (rows: EvalRow[], row: EvalRow, index: number) => void | Promise<void>;
  }
): Promise<EvalRow[]> {
  const total = rows.length;
  const judge = options?.judge ?? createJudge(options?.judgeProvider ?? 'workiq', client, options?.tenantId, options?.judgeAgentId);
  const fallbackJudge = options?.fallbackJudge ?? createDefaultFallbackJudge(judge, client, options?.tenantId, options?.fallbackJudgeProvider);
  const evaluators = options?.evaluators ?? DEFAULT_M365_EVALUATORS;
  const concurrency = Math.max(1, Math.min(options?.concurrency ?? 1, total || 1));
  const delayMs = Math.max(0, options?.delayMs ?? DELAY_MS);
  const throttleGate = createThrottleGate(concurrency);
  let nextIndex = 0;
  let completed = rows.filter(row => row.similarityScore !== undefined).length;

  async function processRow(i: number): Promise<void> {
    const row = rows[i];

    if (row.similarityScore !== undefined) {
      return;
    }

    if (!row.actualAnswer || row.actualAnswer.startsWith('[ERROR:')) {
      row.similarityScore = 0;
      const effectiveEvaluators = resolveRowEvaluators(row, evaluators);
      const primary = firstLlmEvaluator(effectiveEvaluators) ?? 'Similarity';
      row.metrics = mergeMetrics(row.metrics, [{
        name: primary,
        score: 0,
        passed: false,
        reason: !row.actualAnswer ? 'Actual answer is empty.' : 'Actual answer contains an error.',
        provider: judge.provider,
        model: judge.model,
        scale: '0-100',
        threshold: options?.threshold,
      }]);
      setRowError(row, row.actualAnswer ? 'agentRequestFailed' : 'turnSkipped', !row.actualAnswer ? 'Actual answer is empty.' : row.actualAnswer);
      completed++;
      options?.onProgress?.(i + 1, total);
      process.stderr.write(`\rScoring answer ${i + 1}/${total}...`);
      await options?.onRowComplete?.(rows, row, i);
      return;
    }

    process.stderr.write(`\rScoring answer ${i + 1}/${total}...`);

    try {
      const effectiveEvaluators = resolveRowEvaluators(row, evaluators);
      const metrics: MetricResult[] = [];
      const llmEvaluators = effectiveEvaluators.filter(isLlmEvaluator);
      for (const evaluator of llmEvaluators) {
        const scored = await throttleGate.run(() => scoreWithFallback(judge, fallbackJudge, row, evaluator));
        metrics.push(metricFromJudge(scored.score, scored.judge, options?.threshold, evaluator));
      }
      metrics.push(...evaluateDeterministicMetrics(row, effectiveEvaluators, options?.threshold));
      const primaryMetric = metrics.find(metric => metric.name === 'Similarity') ??
        metrics.find(metric => metric.name === 'SemanticSimilarity') ??
        metrics[0];
      row.metrics = mergeMetrics(row.metrics, metrics);
      if (primaryMetric) {
        row.similarityScore = primaryMetric.score ?? 0;
        row.status = deriveRowStatus(row, options?.threshold);
      } else {
        row.similarityScore = undefined;
        row.status = undefined;
      }
    } catch (err: unknown) {
      const message = err instanceof Error ? err.message : String(err);
      process.stderr.write(`\nWarning: Scoring failed for row ${i + 1}: ${message}, setting to 0\n`);
      row.similarityScore = 0;
      row.metrics = mergeMetrics(row.metrics, [{
        name: firstLlmEvaluator(resolveRowEvaluators(row, evaluators)) ?? 'Similarity',
        score: 0,
        passed: false,
        reason: message,
        provider: judge.provider,
        model: judge.model,
        scale: '0-100',
        threshold: options?.threshold,
      }]);
      row.status = deriveRowStatus(row, options?.threshold);
    }

    completed++;
    options?.onProgress?.(completed, total);
    await options?.onRowComplete?.(rows, row, i);

    if (i < total - 1 && delayMs > 0) {
      await delay(delayMs);
    }
  }

  async function worker(): Promise<void> {
    while (nextIndex < total) {
      const current = nextIndex++;
      await processRow(current);
    }
  }

  await Promise.all(Array.from({ length: concurrency }, () => worker()));

  process.stderr.write('\n');
  return rows;
}

function createDefaultFallbackJudge(
  primary: Judge,
  client: WorkIQClient,
  tenantId?: string,
  configuredProvider?: 'github-copilot' | 'azure-openai' | 'none',
): Judge | undefined {
  if (primary.provider !== 'workiq') return undefined;
  const disabled = process.env.EVALSCORE_DISABLE_GITHUB_FALLBACK?.toLowerCase();
  if (configuredProvider === 'none' || disabled === '1' || disabled === 'true' || disabled === 'yes') return undefined;
  const envProvider = process.env.EVALSCORE_FALLBACK_JUDGE_PROVIDER as 'github-copilot' | 'azure-openai' | 'none' | undefined;
  const provider = configuredProvider ?? (envProvider === 'azure-openai' || envProvider === 'github-copilot' ? envProvider : 'github-copilot');
  return createJudge(provider, client, tenantId);
}

async function scoreWithFallback(
  judge: Judge,
  fallbackJudge: Judge | undefined,
  row: EvalRow,
  evaluator: EvaluatorName,
): Promise<{ score: Awaited<ReturnType<Judge['score']>>; judge: Judge }> {
  try {
    return { score: await judge.score(row, evaluator), judge };
  } catch (err) {
    if (!fallbackJudge || !isFallbackEligible(err)) throw err;
    const message = err instanceof Error ? err.message : String(err);
    process.stderr.write(`\nWarning: ${judge.provider} judge failed for ${evaluator}; falling back to ${fallbackJudge.provider}: ${message}\n`);
    const fallbackScore = await fallbackJudge.score(row, evaluator);
    return {
      score: {
        ...fallbackScore,
        reason: fallbackScore.reason
          ? `Fallback from ${judge.provider} due to: ${message}. ${fallbackScore.reason}`
          : `Fallback from ${judge.provider} due to: ${message}.`,
      },
      judge: fallbackJudge,
    };
  }
}

function isFallbackEligible(err: unknown): boolean {
  const message = err instanceof Error ? err.message.toLowerCase() : String(err).toLowerCase();
  return (
    message.includes('timed out') ||
    message.includes('timeout') ||
    message.includes('could not parse score') ||
    message.includes('ask_work_iq') ||
    message.includes('mcp') ||
    message.includes('429') ||
    message.includes('rate limit') ||
    message.includes('throttl') ||
    message.includes('temporarily unavailable') ||
    message.includes('503') ||
    message.includes('502') ||
    message.includes('504')
  );
}

export function calculateScoringResult(rows: EvalRow[], passThreshold: number): ScoringResult {
  const scores = rows
    .map(row => row.similarityScore)
    .filter((score): score is number => score !== undefined);
  const totalQuestions = scores.length;
  const sum = scores.reduce((acc, s) => acc + s, 0);
  const averageScore = totalQuestions > 0 ? Math.round((sum / totalQuestions) * 10) / 10 : 0;
  const minScore = totalQuestions > 0 ? Math.min(...scores) : 0;
  const maxScore = totalQuestions > 0 ? Math.max(...scores) : 0;
  const passCount = scores.filter(s => s >= passThreshold).length;
  const failCount = totalQuestions - passCount;

  // Assertion statistics
  let totalAssertions = 0;
  let assertionsPassed = 0;
  let assertionsFailed = 0;
  for (const row of rows) {
    if (row.assertionResults) {
      totalAssertions += row.assertionResults.length;
      assertionsPassed += row.assertionResults.filter(r => r.passed).length;
      assertionsFailed += row.assertionResults.filter(r => !r.passed).length;
    }
  }

  return {
    totalQuestions,
    averageScore,
    minScore,
    maxScore,
    passCount,
    failCount,
    passThreshold,
    totalAssertions,
    assertionsPassed,
    assertionsFailed,
  };
}

export function parseEvaluators(value?: string): EvaluatorName[] {
  if (!value) return DEFAULT_M365_EVALUATORS;
  const all: EvaluatorName[] = [
    'Similarity',
    'SemanticSimilarity',
    'Relevance',
    'Coherence',
    'Groundedness',
    'Citations',
    'ExactMatch',
    'PartialMatch',
    'EvalGenAssertions',
  ];
  const aliases = new Map<string, EvaluatorName>(all.map(name => [name.toLowerCase(), name]));
  aliases.set('semantic', 'Similarity');
  aliases.set('semanticsimilarity', 'Similarity');
  const parts = value.split(',').map(part => part.trim()).filter(Boolean);
  const names: EvaluatorName[] = [];
  for (const part of parts) {
    if (part.toLowerCase() === 'all') return all;
    const name = aliases.get(part.toLowerCase());
    if (!name) {
      throw new Error(`Unsupported evaluator "${part}". Supported evaluators: ${all.join(', ')}, all`);
    }
    if (!names.includes(name)) names.push(name);
  }
  return names.length > 0 ? names : DEFAULT_M365_EVALUATORS;
}

function evaluateDeterministicMetrics(row: EvalRow, evaluators: EvaluatorName[], threshold?: number): MetricResult[] {
  const metrics: MetricResult[] = [];
  const actual = normalize(row.actualAnswer);
  const expected = normalize(row.expectedAnswer);

  if (evaluators.includes('ExactMatch')) {
    const passed = actual === expected;
    metrics.push({
      name: 'ExactMatch',
      score: passed ? 100 : 0,
      passed,
      reason: passed ? 'Actual answer exactly matches expected answer after normalization.' : 'Actual answer does not exactly match expected answer.',
      provider: 'deterministic',
      scale: '0-100',
      threshold,
    });
  }

  if (evaluators.includes('PartialMatch')) {
    const passed = expected.length > 0 && (actual.includes(expected) || expected.includes(actual));
    metrics.push({
      name: 'PartialMatch',
      score: passed ? 100 : 0,
      passed,
      reason: passed ? 'Actual and expected answers partially overlap.' : 'Actual and expected answers do not partially overlap.',
      provider: 'deterministic',
      scale: '0-100',
      threshold,
    });
  }

  if (evaluators.includes('Citations')) {
    const hasCitation = Boolean(row.citations?.length) ||
      (row.sourceLocation ? row.actualAnswer.toLowerCase().includes(row.sourceLocation.toLowerCase()) : false);
    metrics.push({
      name: 'Citations',
      score: hasCitation ? 100 : 0,
      passed: hasCitation,
      reason: hasCitation ? 'At least one citation/source reference was detected.' : 'No citation/source reference was detected.',
      provider: 'deterministic',
      scale: '0-100',
      threshold,
    });
  }

  if (evaluators.includes('EvalGenAssertions') && row.assertions?.length) {
    row.assertionResults = evaluateRowAssertions(row);
    const passedCount = row.assertionResults.filter(result => result.passed).length;
    const score = Math.round((passedCount / row.assertionResults.length) * 100);
    metrics.push({
      name: 'EvalGenAssertions',
      score,
      passed: passedCount === row.assertionResults.length,
      reason: `${passedCount}/${row.assertionResults.length} assertions passed.`,
      provider: 'deterministic',
      scale: '0-100',
      threshold,
    });
  }

  return metrics;
}

function mergeMetrics(existing: MetricResult[] | undefined, next: MetricResult[]): MetricResult[] {
  const merged = new Map<EvaluatorName, MetricResult>();
  for (const metric of existing ?? []) merged.set(metric.name, metric);
  for (const metric of next) merged.set(metric.name, metric);
  return Array.from(merged.values());
}

function normalize(value: string): string {
  return value.trim().replace(/\s+/g, ' ').toLowerCase();
}

function isLlmEvaluator(evaluator: EvaluatorName): boolean {
  return evaluator === 'Similarity' ||
    evaluator === 'SemanticSimilarity' ||
    evaluator === 'Relevance' ||
    evaluator === 'Coherence' ||
    evaluator === 'Groundedness';
}

function firstLlmEvaluator(evaluators: EvaluatorName[]): EvaluatorName | undefined {
  return evaluators.find(isLlmEvaluator);
}

File: eval-ui-winui3/src/EvalToolkit.EvalScore/Scoring/Scorer.cs
using EvalToolkit.Core.Concurrency;
using EvalToolkit.EvalScore.EvalDocument;
using EvalToolkit.EvalScore.Judges;
using EvalToolkit.EvalScore.Models;
using EvalToolkit.WorkIQ;

namespace EvalToolkit.EvalScore.Scoring;

/// <summary>
/// Scoring pass: dispatches LLM evaluators to a judge (with optional
/// fallback) and computes deterministic metrics, mutating every row's
/// <see cref="EvalRow.Metrics"/>, <see cref="EvalRow.SimilarityScore"/>,
/// <see cref="EvalRow.Status"/>, and (on error) <see cref="EvalRow.Error"/>.
/// Mirrors TS <c>scoreAnswers</c> in <c>eval-score/node/src/scorer.ts</c>.
///
/// <para>Behavior:
/// <list type="bullet">
///   <item>Rows that already have a <see cref="EvalRow.SimilarityScore"/>
///     are skipped (resumability).</item>
///   <item>Rows with no actual answer or an <c>[ERROR:</c>-prefixed
///     actual answer get a zero-score primary metric and the appropriate
///     <see cref="EvalErrorCode"/> attached, without calling any judge.</item>
///   <item>For each remaining row, every LLM evaluator (filtered via
///     <see cref="IsLlmEvaluator(EvaluatorName)"/>) runs through the
///     judge inside a <see cref="ThrottleGate"/>; on fallback-eligible
///     failure the secondary judge takes over with a "Fallback from
///     {primary} due to: {message}." reason prefix preserved.</item>
///   <item>Deterministic metrics (ExactMatch, PartialMatch, Citations,
///     EvalGenAssertions) are appended after LLM metrics.</item>
///   <item>Primary metric is <see cref="EvaluatorName.Similarity"/> if
///     present, else <see cref="EvaluatorName.SemanticSimilarity"/>, else
///     the first metric. The primary metric's score becomes
///     <see cref="EvalRow.SimilarityScore"/>.</item>
///   <item>A try/catch around the entire row's scoring records a single
///     zero-score metric on uncaught exceptions (matches TS warning path).</item>
///   <item>Concurrency is enforced both at the worker-pool level
///     (<see cref="JobScheduler{TRow}"/>) AND inside the row via
///     <see cref="ThrottleGate"/> — the two-part contract from the
///     concurrency-port slice.</item>
/// </list></para>
///
/// <para><b>Scoring builds single-row jobs</b> (no thread grouping),
/// so JobScheduler workers pull rows in original-row FIFO order to
/// match TS <c>scoreAnswers</c>'s <c>nextIndex++</c> worker pattern.
/// Thread grouping in <c>EvaluationJobBuilder</c> is for the evaluator
/// (where conversation chaining matters), not the scorer.</para>
/// </summary>
public static class Scorer
{
    public static async Task<IList<EvalRow>> ScoreAnswersAsync(
        IList<EvalRow> rows,
        IWorkIQClient? client,
        ScoreOptions? options = null,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(rows);
        options ??= new ScoreOptions();

        int total = rows.Count;
        if (total == 0)
        {
            return rows;
        }

        JudgeProvider primaryProvider = options.JudgeProvider ?? JudgeProvider.WorkIq;
        IJudge judge = options.Judge ?? JudgeFactory.Create(primaryProvider, client, options.TenantId, options.JudgeAgentId);
        IJudge? fallbackJudge = options.FallbackJudge ?? FallbackJudgeBuilder.Build(
            judge, client, options.TenantId, options.FallbackJudgeProvider, options.DisableFallbackJudge);

        IReadOnlyList<EvaluatorName> evaluators = options.Evaluators ?? EvalRowHelpers.DefaultM365Evaluators;

        int requestedConcurrency = options.Concurrency ?? 1;
        if (requestedConcurrency < 1)
        {
            requestedConcurrency = 1;
        }
        if (requestedConcurrency > total)
        {
            requestedConcurrency = total;
        }
        if (requestedConcurrency > ThrottleGate.HardCap)
        {
            requestedConcurrency = ThrottleGate.HardCap;
        }

        int delayMs = options.DelayMs ?? JobSchedulerDefaults.DefaultDelayMs;
        if (delayMs < 0)
        {
            delayMs = 0;
        }

        int completed = 0;
        foreach (EvalRow row in rows)
        {
            if (row.SimilarityScore.HasValue)
            {
                completed++;
            }
        }

        // TS scoreAnswers processes rows in original index order via a single
        // shared FIFO counter (nextIndex++). It does NOT group by thread.
        // Build single-row jobs so the JobScheduler preserves original row
        // order; thread grouping in EvaluationJobBuilder is for the evaluator
        // (where conversation chaining matters), not the scorer.
        IReadOnlyList<EvaluationJob<EvalRow>> jobs = EvaluationJobBuilder.Build(
            (IReadOnlyList<EvalRow>)rows,
            (row, idx) => new RowKey(TurnIndex: null, ThreadId: row.ThreadId, ItemId: row.Id, ItemIndex: row.ItemIndex));

        using var gate = new ThrottleGate(requestedConcurrency);
        var scheduler = new JobScheduler<EvalRow>(rows.ToArray(), jobs);
        await scheduler.RunAsync(
            async (rowIndex, _, ct) =>
            {
                EvalRow row = rows[rowIndex];
                // TS returns early for pre-scored rows BEFORE incrementing
                // `completed` or firing onProgress/onRowComplete. The initial
                // seed above already counted these rows.
                if (row.SimilarityScore.HasValue)
                {
                    return null;
                }
                await ProcessRowAsync(row, rowIndex, rows, judge, fallbackJudge, evaluators, options, gate, ct).ConfigureAwait(false);
                int doneNow = Interlocked.Increment(ref completed);
                options.OnProgress?.Invoke(doneNow, total);
                if (options.OnRowCompleteAsync is not null)
                {
                    await options.OnRowCompleteAsync(rows.ToArray(), row, rowIndex, ct).ConfigureAwait(false);
                }
                return null; // scorer never threads conversation ids
            },
            new JobSchedulerOptions
            {
                Concurrency = requestedConcurrency,
                DelayMs = delayMs,
            },
            cancellationToken).ConfigureAwait(false);

        return rows;
    }

    private static async Task ProcessRowAsync(
        EvalRow row,
        int rowIndex,
        IList<EvalRow> allRows,
        IJudge judge,
        IJudge? fallbackJudge,
        IReadOnlyList<EvaluatorName> defaultEvaluators,
        ScoreOptions options,
        ThrottleGate gate,
        CancellationToken cancellationToken)
    {
        _ = rowIndex;
        _ = allRows;
        // Pre-scored short-circuit is enforced by the scheduler callback so
        // skipped rows don't fire onProgress/onRowComplete (TS scoreAnswers
        // returns before increment/callbacks at scorer.ts line 45-47).

        IReadOnlyList<EvaluatorName> effectiveEvaluators = EvalRowHelpers.ResolveRowEvaluators(row, defaultEvaluators);

        if (string.IsNullOrEmpty(row.ActualAnswer) || row.ActualAnswer.StartsWith("[ERROR:", StringComparison.Ordinal))
        {
            row.SimilarityScore = 0;
            EvaluatorName primary = FirstLlmEvaluator(effectiveEvaluators) ?? EvaluatorName.Similarity;
            var metric = new MetricResult
            {
                Name = primary,
                Score = 0,
                Passed = false,
                Reason = string.IsNullOrEmpty(row.ActualAnswer)
                    ? "Actual answer is empty."
                    : "Actual answer contains an error.",
                Provider = MetricProviders.FromJudge(judge.Provider),
                Model = judge.Model,
                Scale = MetricScale.ZeroToOneHundred,
                Threshold = options.Threshold,
            };
            row.Metrics = MergeMetrics(row.Metrics, new[] { metric });
            EvalRowHelpers.SetRowError(
                row,
                string.IsNullOrEmpty(row.ActualAnswer) ? EvalErrorCode.TurnSkipped : EvalErrorCode.AgentRequestFailed,
                string.IsNullOrEmpty(row.ActualAnswer) ? "Actual answer is empty." : row.ActualAnswer);
            return;
        }

        try
        {
            var metrics = new List<MetricResult>();
            foreach (EvaluatorName evaluator in effectiveEvaluators.Where(IsLlmEvaluator))
            {
                (JudgeScore scored, IJudge winningJudge) = await gate.RunAsync(
                    () => ScoreWithFallbackAsync(judge, fallbackJudge, row, evaluator, cancellationToken),
                    cancellationToken).ConfigureAwait(false);
                metrics.Add(MetricBuilder.MetricFromJudge(scored, winningJudge, options.Threshold, evaluator));
            }
            metrics.AddRange(DeterministicEvaluators.Evaluate(row, effectiveEvaluators, options.Threshold));

            MetricResult? primaryMetric =
                metrics.FirstOrDefault(m => m.Name == EvaluatorName.Similarity)
                ?? metrics.FirstOrDefault(m => m.Name == EvaluatorName.SemanticSimilarity)
                ?? metrics.FirstOrDefault();

            row.Metrics = MergeMetrics(row.Metrics, metrics);
            if (primaryMetric is not null)
            {
                row.SimilarityScore = primaryMetric.Score ?? 0;
                row.Status = EvalRowHelpers.DeriveRowStatus(row, (int)Math.Round(options.Threshold ?? 70));
            }
            else
            {
                row.SimilarityScore = null;
                row.Status = null;
            }
        }
        catch (OperationCanceledException)
        {
            throw;
        }
        catch (Exception ex)
        {
            string message = ex.Message;
            row.SimilarityScore = 0;
            IReadOnlyList<EvaluatorName> resolved = EvalRowHelpers.ResolveRowEvaluators(row, defaultEvaluators);
            EvaluatorName primary = FirstLlmEvaluator(resolved) ?? EvaluatorName.Similarity;
            var fallbackMetric = new MetricResult
            {
                Name = primary,
                Score = 0,
                Passed = false,
                Reason = message,
                Provider = MetricProviders.FromJudge(judge.Provider),
                Model = judge.Model,
                Scale = MetricScale.ZeroToOneHundred,
                Threshold = options.Threshold,
            };
            row.Metrics = MergeMetrics(row.Metrics, new[] { fallbackMetric });
            row.Status = EvalRowHelpers.DeriveRowStatus(row, (int)Math.Round(options.Threshold ?? 70));
        }
    }

    private static async Task<(JudgeScore Score, IJudge Judge)> ScoreWithFallbackAsync(
        IJudge primary,
        IJudge? fallback,
        EvalRow row,
        EvaluatorName evaluator,
        CancellationToken cancellationToken)
    {
        try
        {
            JudgeScore primaryScore = await primary.ScoreAsync(row, evaluator, cancellationToken).ConfigureAwait(false);
            return (primaryScore, primary);
        }
        catch (OperationCanceledException)
        {
            throw;
        }
        catch (Exception ex)
        {
            if (fallback is null || !FallbackClassifier.IsEligible(ex))
            {
                throw;
            }
            JudgeScore fbScore = await fallback.ScoreAsync(row, evaluator, cancellationToken).ConfigureAwait(false);
            string prefix = $"Fallback from {primary.Provider.ToWireString()} due to: {ex.Message}.";
            string combinedReason = string.IsNullOrEmpty(fbScore.Reason)
                ? prefix
                : $"{prefix} {fbScore.Reason}";
            var wrapped = new JudgeScore(fbScore.Score, combinedReason, fbScore.Model);
            return (wrapped, fallback);
        }
    }

    private static List<MetricResult> MergeMetrics(IList<MetricResult>? existing, IReadOnlyList<MetricResult> incoming)
    {
        // TS: merge by evaluator name, last-write-wins, ordered by
        // (existing-then-incoming) insertion order.
        var merged = new Dictionary<EvaluatorName, MetricResult>();
        var order = new List<EvaluatorName>();
        if (existing is not null)
        {
            foreach (MetricResult m in existing)
            {
                if (!merged.ContainsKey(m.Name))
                {
                    order.Add(m.Name);
                }
                merged[m.Name] = m;
            }
        }
        foreach (MetricResult m in incoming)
        {
            if (!merged.ContainsKey(m.Name))
            {
                order.Add(m.Name);
            }
            merged[m.Name] = m;
        }
        var result = new List<MetricResult>(order.Count);
        foreach (EvaluatorName key in order)
        {
            result.Add(merged[key]);
        }
        return result;
    }

    private static bool IsLlmEvaluator(EvaluatorName evaluator) => evaluator is
        EvaluatorName.Similarity or
        EvaluatorName.SemanticSimilarity or
        EvaluatorName.Relevance or
        EvaluatorName.Coherence or
        EvaluatorName.Groundedness;

    private static EvaluatorName? FirstLlmEvaluator(IReadOnlyList<EvaluatorName> evaluators)
    {
        foreach (EvaluatorName e in evaluators)
        {
            if (IsLlmEvaluator(e))
            {
                return e;
            }
        }
        return null;
    }
}
